sources/ghostdog/src/infiniband.rs

```diff
@@ -0,0 +1,197 @@
+    /// Port GUID is a u64 that is read in as a : formated string but written out in a form such as 0x
+    pub(crate) port_guid: Guid,
+}
+
+impl InfinibandPort {
+    /// is_sm_enabled checks the specific capability bit for Subnet Management means its enabled
+    // The actual bit means `isSMDisabled`, so check if its not zero
+    pub(crate) fn is_sm_enabled(&self) -> bool {
+        let bit_position: u32 = 10;
+        let mask: u32 = 1 << bit_position;
+        if !(&self.capability_mask & mask) != 0 {
+            return true;
+        }
+        false
+    }
+}
+
+// Given a device in sysfs, iterate over the known paths where ports are defined
+// Iterate over the ports and read in the capability mask and Port GUID for each port
+pub(crate) fn find_ports_for_device(device: OsString) -> Result<Vec<InfinibandPort>> {
+    let mut found_ports: Vec<InfinibandPort> = vec![];
+    let ports_path = Path::new(SYS_CLASS_DIR)
+        .join(SW_MNG_CLASS)
+        .join(device)
+        .join("ports");
+
+    let port_directories = fs::read_dir(ports_path).context(error::InfinibandSysDevicesSnafu)?;
+    for port in port_directories {
+        let mut capability_mask: String = "".to_string();
+        let mut first_guid: String = "".to_string();
+
+        let port = port.context(error::InfinibandSysDevicesSnafu)?;
+        let capability_mask_path = port.path().join("cap_mask");
+        if capability_mask_path.exists() {
+            capability_mask = std::fs::read_to_string(capability_mask_path.as_path()).context(
+                error::ReadFileSnafu {
+                    path: capability_mask_path.as_path(),
+                },
+            )?;
+        }
+        let first_guid_path = port.path().join("gids").join("0");
+        if first_guid_path.exists() {
+            first_guid = std::fs::read_to_string(first_guid_path.as_path()).context(
+                error::ReadFileSnafu {
+                    path: first_guid_path.as_path(),
+                },
+            )?;
+        }
+        // If a capability mask and port guid have been read, create the Infiniband Port
+        if !capability_mask.is_empty() && !first_guid.is_empty() {
+            found_ports.push(InfinibandPort {
+                capability_mask: CapabilityMask::from_str(capability_mask.as_str())?,
+                port_guid: Guid::from_str(first_guid.as_str())?,
+            });
+        }
+    }
+    Ok(found_ports)
+}
+
+/// Guid struct to help with custom FromStr and ToString
+#[derive(Clone, Debug)]
+pub(crate) struct Guid(String);
+
+impl FromStr for Guid {
+    type Err = error::Error;
+    /// String is expected to look like "fe80:0000:0000:0000:e09d:7303:003f:3bf8"
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let potential_guid = s.trim().to_string();
+        ensure!(
+            potential_guid.starts_with("fe80"),
+            error::InvalidGuidStringSnafu {
+                guid: potential_guid
+            }
+        );
```

Comment:
What is the source of this `fe80` check?

It'd be cleaner to do this via regex; right now we just quietly return the unmodified value in `to_config_string` but it might not be in the correct format if we don't have all 8 parts.